From 9f80c2adee7b0bf1956603580d1a1e25c022aa67 Mon Sep 17 00:00:00 2001
From: Billy <xiaobing.feng@windriver.com>
Date: Fri, 28 Jun 2013 17:00:07 +0800
Subject: Volume: Enable "Volume mute" key on HSW

Fix:
Android doesn't use master stream by default (config_useMasterVolume=false),
it causes AudioManager to not handle VOLUME_MUTE event. Modify AudioManager
to deal with this event, which makes "Volume mute" key work.

There are 3 kinds of behavior for "Volume mute" key: "Mute master stream",
"Mute activity stream", and "Mute all affected streams". We select the
first one as its behavior. The reasons are:

1. For "Mute activity stream", other streams still work. Once the activity stream
switches out, it is not easy to unmute it for user, and this behavior is different with
other desktop system usage habit, like windows.
2. For "Mute all affected streams", if the streams have different original status, some are
muted, the others are not muted. How can we define the behavior once "Volume mute" key
is pressed down, mute the unmuted streams or unmute the muted streams?

The volume mute implementation uses "SilentMode" (RingerMode)
to mute and "NormalMode" to unmute. SilentMode mutes music, video, games, system sounds,
other media and notification sounds.

Category: Device Enablement
Domain: IO
Upstream-Candidate: yes

Issue: AXIA-2773
Change-Id: I87179ab1ef58ca9973b6f4421039ec492fb417a7
Signed-off-by: Xiaobing Feng <xiaobing.feng@windriver.com>
---
 media/java/android/media/AudioManager.java |   14 +++++++++++++-
 1 files changed, 13 insertions(+), 1 deletions(-)

diff --git a/media/java/android/media/AudioManager.java b/media/java/android/media/AudioManager.java
index d652cae..585cd63 100644
--- a/media/java/android/media/AudioManager.java
+++ b/media/java/android/media/AudioManager.java
@@ -16,6 +16,7 @@
 
 package android.media;
 
+import static android.media.AudioManager.RINGER_MODE_SILENT;
 import android.Manifest;
 import android.annotation.SdkConstant;
 import android.annotation.SdkConstant.SdkConstantType;
@@ -536,7 +537,18 @@ public class AudioManager {
                     if (mUseMasterVolume) {
                         setMasterMute(!isMasterMute());
                     } else {
-                        // TODO: Actually handle MUTE.
+                        /*
+                         * Using silent mode for mute implementation.
+                         * Affected streams include music,video,games,media,
+                         * system and notification sounds.
+                         */
+                        boolean muteOn = (getRingerMode() == RINGER_MODE_SILENT);
+
+                        if (muteOn) {   // Toggle mute
+                            setRingerMode(RINGER_MODE_NORMAL);
+                        } else {
+                            setRingerMode(RINGER_MODE_SILENT);
+                        }
                     }
                 }
                 break;
-- 
1.7.4.1

